src/D2L.Bmx/OktaAuthenticator.cs

```diff
@@ -189,6 +148,65 @@ The provided Okta user '{providedLogin}' does not match the system configured pa
 		return oktaAuthenticatedClient;
 	}
 
+	private async Task<string?> GetSessionIdFromBrowserAsync( string browserPath, Uri orgUrl ) {
+		await using var browser = await browserLauncher.LaunchAsync( browserPath );
+
+		var sessionIdTcs = new TaskCompletionSource<string?>( TaskCreationOptions.RunContinuationsAsynchronously );
+
+		// cancel if the total time exceeds 15 seconds, including all page loads and retries
+		using var cancellationTokenSource = new CancellationTokenSource( TimeSpan.FromSeconds( 15 ) );
+		cancellationTokenSource.Token.Register( () => sessionIdTcs.TrySetCanceled() );
+
+		// cancel if we're stuck on a single page for 3 seconds
+		using var pageTimer = new System.Timers.Timer( TimeSpan.FromSeconds( 3 ) ) { AutoReset = false };
+		pageTimer.Elapsed += ( _, _ ) => cancellationTokenSource.Cancel();
+		pageTimer.Start();
+
+		using var page = await browser.NewPageAsync().WaitAsync( cancellationTokenSource.Token );
+		int attempt = 1;
+
+		page.Load += ( _, _ ) => _ = OnPageLoadAsync();
+		await page.GoToAsync( orgUrl.AbsoluteUri ).WaitAsync( cancellationTokenSource.Token );
+		return await sessionIdTcs.Task;
+
+		async Task OnPageLoadAsync() {
+			// reset the 3-sec per-page timer on every page load
+			lock( pageTimer ) {
+				pageTimer.Stop();
+				pageTimer.Start();
```

Comment:
Don't follow why ``.stop`` needs to be called here?

Reply:
Calling `.Start()` on an already started timer does nothing - it will still fire at the originally scheduled time.
We want to restart the timer here, so `.Stop()` must be called first to reset it.